neqo-http3/src/connection.rs

```diff
@@ -613,23 +670,83 @@ impl<E: Http3Events + Default, T: Http3Transaction, H: Http3Handler<E, T>>
         Ok(())
     }
 
-    fn handle_settings(&mut self, s: &[(HSettingType, u64)]) -> Res<()> {
-        qinfo!([self], "Handle SETTINGS frame.");
+    fn set_qpack_settings(&mut self, s: &[(HSettingType, u64)]) -> Res<()> {
         for (t, v) in s {
             qinfo!([self], " {:?} = {:?}", t, v);
             match t {
-                HSettingType::MaxHeaderListSize => {
-                    self.max_header_list_size = *v;
-                }
                 HSettingType::MaxTableSize => self.qpack_encoder.set_max_capacity(*v)?,
                 HSettingType::BlockedStreams => self.qpack_encoder.set_max_blocked_streams(*v)?,
-
                 _ => {}
             }
         }
         Ok(())
     }
 
+    fn handle_settings(&mut self, s: &[(HSettingType, u64)]) -> Res<()> {
+        qinfo!([self], "Handle SETTINGS frame.");
+        match &self.settings_state {
+            Http3RemoteSettingsState::NotReceived => {
+                self.set_qpack_settings(s)?;
+                self.settings_state = Http3RemoteSettingsState::Received(s.to_vec());
+                Ok(())
+            }
+            Http3RemoteSettingsState::ZeroRtt(settings) => {
```

Comment:
This function a little too much knowledge about settings.  Some of that could be moved to a wrapper.  I think that this would be much easier if you had a class that wrapped the finding of values with defaults.   

Before you ask, maximum header list size can default to 2^62 if you are concerned about "unlimited"; the effect is the same.

Then the calculation is easier: you find the value, check if it is bigger.  Then this code is:

```rust
Http3RemoteSettingsState::ZeroRtt(old_settings) => { 
  let mut qpack_changed = false;
  for st in &[  ... ] {
    let old_value = old_settings.get(s);
    let new_value = new_settings.get(s);
    let diff = i64::try_from(new_settings.get(s)).unwrap() - i64::try_from(old_value).unwrap();
    if old_value == new_value { continue; }
    if old_value > new_value { error!(); }
    match s {
      HSettingsType::MaxTableCapacity => {
        if old_value != 0 { error!(); }
        qpack_changed = true;
      },
      HSettingsType::BlockedStreams => qpack_changed = true,
      _ => (),
    }
  }
  if qpack_changed { self.qpack.update_settings(); }
  Ok(())
```